src/NHibernate/Linq/Visitors/HqlGeneratorExpressionVisitor.cs

```diff
@@ -579,8 +579,12 @@ protected HqlTreeNode VisitConstantExpression(ConstantExpression expression)
 			if (_parameters.ConstantToParameterMap.TryGetValue(expression, out namedParameter))
 			{
 				_parameters.RequiredHqlParameters.Add(new NamedParameterDescriptor(namedParameter.Name, null, false));
+				var parameter = _hqlTreeBuilder.Parameter(namedParameter.Name).AsExpression();
 
-				return _hqlTreeBuilder.Parameter(namedParameter.Name).AsExpression();
+				// SQLite driver binds decimal parameters to text, which can cause unexpected results in arithmetic operations.
+				return expression.Type.UnwrapIfNullable() == typeof(decimal) && _parameters.SessionFactory.Dialect.IsDecimalStoredAsFloatingPointNumber
```

Comment:
Better swap those checks... Also maybe make `IsDecimalStoredAsFloatingPointNumber` public and use in tests instead of Dialect checks.

Reply:
Done.